pkg/util/cache/cache.go

```diff
@@ -284,25 +288,32 @@ func (c *lruCache[K, V]) DoWait(key K, timeout time.Duration, doer func(V) error
 		}()
 		select {
 		case <-c:
-			return false // completed normally
+			return nil // completed normally
 		case <-time.After(timeout):
-			return true // timed out
+			return ErrTimeOut // timed out
+		case <-ctx.Done():
+			return ctx.Err() // context timeout
 		}
 	}
 
 	var ele *list.Element
 	start := time.Now()
+	log := log.Ctx(ctx).With(zap.Any("key", key))
 	for {
-		item, missing, err := c.getAndPin(key)
+		item, missing, err := c.getAndPin(ctx, key)
 		if err == nil {
 			if ele != nil {
 				c.rwlock.Lock()
 				c.waitQueue.Remove(ele)
 				c.rwlock.Unlock()
 			}
 			defer c.Unpin(key)
-			return missing, doer(item.value)
-		} else if err != ErrNotEnoughSpace {
+			return missing, doer(ctx, item.value)
+		} else if err == ErrNotEnoughSpace {
+			log.Warn("Failed to get disk cache for segment, wait and try again")
+		} else if err == merr.ErrSegmentLoadFailed {
+			log.Warn("Failed to load segment, wait and try again")
+		} else if err != nil {
 			return true, err
```

Comment:
When loading failed, make cache to try again rather than simply failed. Maybe we need to enrich error returned here

Reply:
Please be noted that the retry here is endless until succeeded or timeout. If the failure is determined (e.g., file broken, file not found, etc.), there would be likely a dead loop.

Reply:
changed